openmc/capi/filter.py

```diff
@@ -77,7 +102,10 @@ def id(self, filter_id):
         _dll.openmc_filter_set_id(self._index, filter_id)
 
 
-class EnergyFilterView(FilterView):
+class EnergyFilter(Filter):
+    def __new__(cls, bins=None, uid=None, new=True, index=None):
+        return super(EnergyFilter, cls).__new__(cls, b'energy', uid, new, index)
```

Comment:
There is a lot of duplication in these `__new__` functions.  What do you think of using something like the `FilterMeta` class with its `short_name` attribute to automatically fill the `filter_type` field of `__new__`?

Reply:
Yeah, it's possible that some of the duplication could be handled with a metaclass. I was going to wait until we go Py3-only and try to use a class decorator instead, but if you'd prefer I can look at the metaclass approach now.

Reply:
What decorator would you use?

Reply:
One that I wrote myself.

Reply:
Sure, but what would the decorator look like?  Would it be something where you explicitly pass a name like this?
```python
@filter_name(b'cell')
class CellFilter(Filter):
    pass
```

Or would it use some introspection for
```python
@filter_type
class CellFilter(Filter):
    pass
```

Reply:
I haven't gotten that far. The other thing I was thinking about was how `__new__` in `Filter`, `Tally`, `Material`, etc. is essentially the same and could potentially be replaced by a class decorator that adds a `__new__` method with the appropriate logic/customization.

Reply:
Ooh, I'd be interested to see how decorators help us clean up duplication between `Tally.__new__`, `Material.__new__`, etc.  If we find something nice for that case, we might be able to do the same for the MGXS module.

Anyway, I think metaclassing is probably more appropriate for this case.  I think the actual code in the the metaclass' `__new__` method and the decorator will be very similar, and the decorator would require more LoC since it has to be explicitly invoked every time we define a new filter class.

Reply:
Indeed, MGXS is in need of serious cleanup. For now, I might just add a `filter_type` class attribute on each of these classes that is referenced in `Filter.__new__`. [Simple is better than complex](https://www.python.org/dev/peps/pep-0020/).